src/Cardano/Prelude/GHC/Heap/Size.hs

```diff
@@ -0,0 +1,66 @@
+{-# LANGUAGE BangPatterns        #-}
+{-# LANGUAGE MagicHash           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UnboxedTuples       #-}
+
+module Cardano.Prelude.GHC.Heap.Size (
+    computeHeapSize
+  ) where
+
+import Cardano.Prelude.Base hiding (Any)
+
+import GHC.Exts
+import GHC.Exts.Heap
+import System.Mem (performGC)
+
+computeHeapSize :: a -> IO Word64
+computeHeapSize = \a -> do
+    performGC
+    go 0 [] [Box $ unsafeCoerce# a]
+  where
+    go :: Word64 -> [Box] -> [Box] -> IO Word64
+    go !acc _    []     = return acc
+    go !acc done (b:bs) = do
+        alreadyDone <- boxElem b done
+        if alreadyDone then
+          go acc done bs
+        else do
+          let !(rawSize, pointers) = unpackBox b
+          go (acc + rawSize) (b:done) (pointers ++ bs)
+
+-- | (Short-circuited) check if a box is in a list
+boxElem :: Box -> [Box] -> IO Bool
+boxElem b' = go
+  where
+    go :: [Box] -> IO Bool
+    go []     = return False
+    go (b:bs) = do
+        equal <- areBoxesEqual b' b
+        if equal then return True
+                 else go bs
+
+-- | The raw size and nested pointers in a closure
+unpackBox :: Box -> (Word64, [Box])
+unpackBox (Box a) = (rawSize, arrayToList pointers)
+  where
+    _addr    :: Addr#
+    raw      :: ByteArray#
+    pointers :: Array# Any
+    !(# _addr, raw, pointers #) = unpackClosure# a
+
+    rawSize :: Word64
+    !rawSize = fromIntegral $ I# (sizeofByteArray# raw)
+
+-- | Translate array to list
+--
+-- When the resulting list is reduced to WHNF the array is no longer retained.
+arrayToList :: Array# Any -> [Box]
+arrayToList arr = getElems [] [0 .. numElems - 1]
+  where
+    getElems :: [Box] -> [Int] -> [Box]
+    getElems acc []          = acc
+    getElems acc (I# i : is) = case indexArray# arr i of
+                                 (# x #) -> getElems (Box x : acc) is
```

Comment:
Mention that forcing the resulting list to WHNF means we no longer hold on to the array.